DataFormats/TrackerRecHit2D/src/VectorHit.cc

```diff
@@ -0,0 +1,413 @@
+      theChi2(vh.chi2()),
+      theDimension(vh.dimension()),
+      theLowerCluster(vh.lowerClusterRef()),
+      theUpperCluster(vh.upperClusterRef()) {}
+
+VectorHit::VectorHit(const GeomDet& idet,
+                     const LocalPoint& posLower,
+                     const LocalVector& dir,
+                     const AlgebraicSymMatrix& covMatrix,
+                     const double& Chi2,
+                     OmniClusterRef const& lower,
+                     OmniClusterRef const& upper)
+    : BaseTrackerRecHit(idet, trackerHitRTTI::vector),
+      thePosition(posLower),
+      theDirection(dir),
+      theCovMatrix(covMatrix),
+      theChi2(Chi2),
+      theDimension(4),
+      theLowerCluster(lower),
+      theUpperCluster(upper) {}
+
+VectorHit::VectorHit(const GeomDet& idet,
+                     const VectorHit2D& vh2Dzx,
+                     const VectorHit2D& vh2Dzy,
+                     OmniClusterRef const& lower,
+                     OmniClusterRef const& upper)
+    : BaseTrackerRecHit(idet, trackerHitRTTI::vector), theDimension(4), theLowerCluster(lower), theUpperCluster(upper) {
+  thePosition = LocalPoint(vh2Dzx.localPosition().x(), vh2Dzy.localPosition().x(), 0.);
+
+  theDirection = LocalVector(vh2Dzx.localDirection().x(), vh2Dzy.localDirection().x(), 1.);
+
+  //building the cov matrix 4x4 starting from the 2x2
+  AlgebraicSymMatrix22 covMatZX = vh2Dzx.covMatrix();
+  AlgebraicSymMatrix22 covMatZY = vh2Dzy.covMatrix();
+
+  theCovMatrix = AlgebraicSymMatrix(4);
+  theCovMatrix[0][0] = covMatZX[0][0];  // var(dx/dz)
+  theCovMatrix[1][1] = covMatZY[0][0];  // var(dy/dz)
+  theCovMatrix[2][2] = covMatZX[1][1];  // var(x)
+  theCovMatrix[3][3] = covMatZY[1][1];  // var(y)
+  theCovMatrix[0][2] = covMatZX[0][1];  // cov(dx/dz,x)
+  theCovMatrix[1][3] = covMatZY[0][1];  // cov(dy/dz,y)
+
+  theChi2 = vh2Dzx.chi2() + vh2Dzy.chi2();
+}
+
+bool VectorHit::sharesInput(const TrackingRecHit* other, SharedInputType what) const {
+  if (what == all && (geographicalId() != other->geographicalId()))
+    return false;
+
+  if (!sameDetModule(*other))
+    return false;
+
+  if (trackerHitRTTI::isVector(*other)) {
+    const VectorHit* otherVh = static_cast<const VectorHit*>(other);
+    return sharesClusters(*this, *otherVh, what);
+  }
+
+  if (what == all)
+    return false;
+
+  // what about multi???
+  auto const& otherClus = reinterpret_cast<const BaseTrackerRecHit*>(other)->firstClusterRef();
+  return (otherClus == lowerClusterRef()) || (otherClus == upperClusterRef());
+}
+
+bool VectorHit::sharesClusters(VectorHit const& h1, VectorHit const& h2, SharedInputType what) const {
+  bool lower = h1.lowerClusterRef() == h2.lowerClusterRef();
+  bool upper = h1.upperClusterRef() == h2.upperClusterRef();
+
+  return (what == TrackingRecHit::all) ? (lower && upper) : (upper || lower);
+}
+
+void VectorHit::getKfComponents4D(KfComponentsHolder& holder) const {
+  AlgebraicVector4& pars = holder.params<4>();
```

Comment:
@kpedro88 I'm trying to find ways to use theDimension or a different const int member variable for this, but keep running into errors like this: `use of 'this' in a constant expression`

Reply:
You need to declare `theDimension` as `static constexpr` for this to work

Reply:
thanks. Still, I am not sure this will work, as `theDimension` is initialized as 0 in the () constructor. I could probably add an additional variable to take on the function you have in mind.

Reply:
I understand this is how it is currently done, but it is not clear to me that there is any purpose in allowing `theDimension` to take a value other than 4.

Reply:
I tend to agree. I have changed `theDimension` to be a `static constexpr` and used it to replace all those 4s